### PR TITLE
feat(protocol): provided an alternative staking implementation

### PR DESCRIPTION
**Pull request by dantaik**
@adaki2004 please take a look at my implementation which I believe is easier to understand. Feel free to comment and ask more questions.

The tests are current failing:

```
Failing tests:
Encountered 8 failing tests in test/TaikoL1Oracle.t.sol:TaikoL1OracleTest
[FAIL. Reason: ERC20: transfer amount exceeds balance] testOracleProverCanAlwaysOverwriteIfNotSameProof() (gas: 83856)
[FAIL. Reason: ERC20: transfer amount exceeds balance] testOracleProverCannotOverwriteIfSameProof() (gas: 83814)
[FAIL. Reason: ERC20: burn amount exceeds balance] testOracleProverWithSignature() (gas: 348035)
[FAIL. Reason: ERC20: burn amount exceeds balance] test_if_oracle_is_disabled_cooldown_is_still_as_proofRegularCooldown() (gas: 401738)
[FAIL. Reason: ERC20: burn amount exceeds balance] test_if_oracle_proofs_can_be_verified_without_regular_proofs() (gas: 401704)
[FAIL. Reason: ERC20: burn amount exceeds balance] test_if_systemProver_can_prove_but_regular_provers_can_overwrite() (gas: 401705)
[FAIL. Reason: ERC20: burn amount exceeds balance] test_if_there_is_no_system_and_oracle_provers() (gas: 401650)
[FAIL. Reason: ERC20: burn amount exceeds balance] test_that_simple_oracle_prover_cannot_be_verified_only_if_normal_proof_comes_in() (gas: 401771)

Encountered a total of 8 failing tests, 90 tests succeeded
 ELIFECYCLE  Test failed. See above for more details.
```

### CODE REVIEW

**Comment by Brechtpd** [comment]
I find both of the implementations pretty complex honestly. They also don't really seem to optimize for the critical path of efficiently checking for each block who can prove it (which is the thing that should be as cheap as possible). 

Just throwing yet another implementation out there (which maybe doesn't help, but at least another point of view!). The idea below is to just let stakers claim a limited set of slots based on how much they stake. These slots will then be used to select a random prover.

The implementation is less than 100 lines of super simple code that I think does everything that needs to be done:
- Constant gas cost to check who the prover for a block is
- Unlimited amount of stakers supported
- Could change the number of slots on the fly depending on what is needed (but we don't exactly expect a lot of prover so a small amount of slots should be okay)
- The smart contract practically doesn't do anything, and the offchain logic should be very simple

```
contract TestStakeContract {
    uint256 public constant NUM_SLOTS = 128;
    uint256 public constant EXIT_PERIOD = 1 weeks;

    uint public totalStaked;

    struct Staker {
       uint256 amount;
       uint numSlots;
       uint unstakedAt;
    }

    mapping(uint slot => address) slots;
    mapping(address staker => Staker) stakers;

    function assignProver(uint64 blockId, uint32 feePerGas)
        external
        view
        returns (address prover, uint32 rewardPerGas)
    {
        bytes32 rand = keccak256(abi.encode(blockId));
        uint256 slot_idx = uint256(rand) % NUM_SLOTS;
        return (slots[slot_idx], feePerGas);
    }

    function stake(address staker, uint amount)
	    external
    {
        totalStaked += amount;
        stakers[staker].amount += amount;
        stakers[staker].unstakedAt = 0;
    }

    function unstake(address staker)
	    external
    {
        stakers[staker].unstakedAt = block.timestamp;
        totalStaked -= stakers[staker].amount;
    }

    function claimSlot(address staker, uint slotIdx)
	    external
    {
        require(staker != address(0));
        // We only allow claiming slots from other stakers if they have more than their minimum amount
        // of slots claimed. We allow anyone to claim slots to take into rounding errors.
        // TODO: currently this would allow people to battle over these slot, so just let the top staker claim these
        require(isSlotClaimable(slotIdx));
        if (stakers[slots[slotIdx]].numSlots > 0) {
            stakers[slots[slotIdx]].numSlots -= 1;
        }
        slots[slotIdx] = staker;
        stakers[staker].numSlots += 1;
    }

    function slashProver(address staker) 
        external 
    {
        uint amountToSlash = 123456;
        amountToSlash = stakers[staker].amount > amountToSlash ? amountToSlash : stakers[staker].amount;
        stakers[staker].amount -= amountToSlash;
    }

    function withdraw(address staker)
	    public
    {
        require(stakers[staker].unstakedAt + EXIT_PERIOD >= block.timestamp);
        stakers[staker].amount = 0;
    }

    function getNumClaimableSlots(address staker)
	    public
        view
	    returns (uint)
    {
        if (stakers[staker].unstakedAt == 0) {
            return (stakers[staker].amount * NUM_SLOTS) / totalStaked;
        } else {
            return 0;
        }
    }

    function isSlotClaimable(uint slotIdx)
	    public
        view
	    returns (bool)
    {
        address currentStaker = slots[slotIdx];
        if (currentStaker == address(0)) {
            return true;
        } else  {
            return stakers[currentStaker].numSlots > getNumClaimableSlots(currentStaker);
        }
    }

    // HELPER FUNCTION ONLY!!! 
    // ONLY HERE SO STAKERS CAN JUST GO TO ETHERSCAN AND FIND ALL SLOTS THEY COULD CLAIM!!!
    // NOT USED IN THE SMART CONTRACT ITSELF!!!
    function getClaimableSlots()
	    public
        view
	    returns (uint[] memory)
    {
        uint[] memory claimableSlots = new uint[](NUM_SLOTS);
        uint numClaimableSlots = 0;
        for (uint i; i < NUM_SLOTS; i++) {
            if (isSlotClaimable(i)) {
                claimableSlots[numClaimableSlots] = i;
                numClaimableSlots += 1;
            }
        }
        // Overwrite the length
        assembly {
            mstore(claimableSlots, numClaimableSlots)
        }
        return claimableSlots;
    }
}

```

**Comment by dantaik** [comment]
Brecht, your idea is interesting but it lacks a way of gradually changing the feePerGas.

File: packages/protocol/contracts/L1/ProverPool2.sol [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        uint16 currentCapacity;
+    }
+
+    // Make sure we only use one slot
+    struct Staker {
+        uint64 exitRequestedAt;
+        uint32 exitAmount;
+        uint16 maxCapacity;
+        uint8 proverId; // 0 to indicate the staker is not a top prover
+    }
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/ProverPool2.sol:31]: Not really, if a prover is new, we still have this Staker object with a non-zero proverId field, all other fields are exit releated.

File: packages/protocol/contracts/L1/ProverPool2.sol [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+    function stake(
+        uint32 amount,
+        uint16 rewardPerGas,
+        uint16 maxCapacity
+    )
+        external
+        nonReentrant
+    {
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/ProverPool2.sol:161]: simply call stake(0,0,0)

File: packages/protocol/contracts/L1/ProverPool2.sol [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+        }
+
+        prover.stakedAmount = 0;
+        prover.rewardPerGas = 0;
+        prover.currentCapacity = 0;
+
+        emit Exited(addr, staker.exitAmount);
+    }
+
+    function _withdraw(address addr) private returns (bool success) {
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/ProverPool2.sol:307]: _exit_ a prover from the prover pool, withdraw _token_ from the pool.

File: packages/protocol/contracts/L1/ProverPool2.sol [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+            uint32 _additional = amountToSlash - staker.exitAmount;
+            if (prover.stakedAmount > _additional) {
+                provers[staker.proverId - 1].stakedAmount -= _additional;
+            } else {
+                provers[staker.proverId - 1].stakedAmount = 0;
+            }
+        }
+
+        emit Slashed(addr, amountToSlash);
+    }
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/ProverPool2.sol:152]: If an exit is "mature" (old enough), it shall not be considered to be part of the "slashable amount".

File: packages/protocol/contracts/L1/ProverPool2.sol [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+        uint256 z;
+        uint8 idx;
+        unchecked {
+            while (z < r && idx < 32) {
+                z += weights[idx++];
+            }
+            provers[idx].currentCapacity--;
+
+            // Note that prover ID is 1 bigger than its index
+            return (idToProver[idx + 1], provers[idx].rewardPerGas);
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/ProverPool2.sol:101]: I see, but do we need to 'break' there?

**Comment by Brechtpd** [comment]
> Brecht, your idea is interesting but it lacks a way of gradually changing the feePerGas.

Oh yeah the code I shared will surely not have all the bells and whistles of an actual implementation, but I don't think it's a limitation of the method and could easily be added.

**Pull request merged**